### PR TITLE
Fix potential deadlock if `calculateSyncRequest` panics

### PR DESCRIPTION
**Pull request by neilalexander**
As `calculateSyncRequest` requires the lock to be held, but the callers in `processSnapshot` also have a `defer` which tries to take the lock, any scenario in which `calculateSyncRequest` panics will result in a deadlock scenario on the stream lock.

As seen in cases like:
```
goroutine 161529372 [sync.RWMutex.Lock, 299 minutes]:
sync.runtime_SemacquireRWMutex(0x484a7a?, 0x28?, 0xc00419c2f0?)
	runtime/sema.go:87 +0x25
*** CALLER TRIES TO LOCK AGAIN: ***
sync.(*RWMutex).Lock(0x0?)
	sync/rwmutex.go:152 +0x6a
github.com/nats-io/nats-server/v2/server.(*stream).processSnapshot.func2()
	github.com/nats-io/nats-server/v2/server/jetstream_cluster.go:8048 +0x45
*** PANIC CAUSES DEFER IN CALLER: ***
panic({0xabf280?, 0x106a1b0?})
	runtime/panic.go:920 +0x270
*** LOCK TAKEN BEFORE HERE: ***
github.com/nats-io/nats-server/v2/server.(*stream).calculateSyncRequest(0xc005253880?, 0xc011094750, 0xc00449bc00)
	github.com/nats-io/nats-server/v2/server/jetstream_cluster.go:7862 +0x42
github.com/nats-io/nats-server/v2/server.(*stream).processSnapshot(0xc00162b500, 0xc00449bc00)
	github.com/nats-io/nats-server/v2/server/jetstream_cluster.go:8123 +0x75b
github.com/nats-io/nats-server/v2/server.(*jetStream).applyStreamEntries(0xc0001c2500, 0xc00162b500, 0x0?, 0x0)
	github.com/nats-io/nats-server/v2/server/jetstream_cluster.go:3083 +0x1a65
github.com/nats-io/nats-server/v2/server.(*jetStream).monitorStream(0xc0001c2500, 0xc00162b500, 0xc0095610e0, 0x0)
	github.com/nats-io/nats-server/v2/server/jetstream_cluster.go:2406 +0x2811
github.com/nats-io/nats-server/v2/server.(*jetStream).processClusterCreateStream.func1()
	github.com/nats-io/nats-server/v2/server/jetstream_cluster.go:3808 +0x25
github.com/nats-io/nats-server/v2/server.(*Server).startGoRoutine.func1()
	github.com/nats-io/nats-server/v2/server/server.go:3740 +0x32
created by github.com/nats-io/nats-server/v2/server.(*Server).startGoRoutine in goroutine 161529608
	github.com/nats-io/nats-server/v2/server/server.go:3738 +0x145
```

Signed-off-by: Neil Twigg <neil@nats.io>

### CODE REVIEW

File: server/jetstream_cluster.go [hunk 1]
```diff
@@ -7912,7 +7912,13 @@ type streamSyncRequest struct {
 }
 
 // Given a stream state that represents a snapshot, calculate the sync request based on our current state.
+// Stream lock must be held.
 func (mset *stream) calculateSyncRequest(state *StreamState, snap *StreamReplicatedState) *streamSyncRequest {
+	// Shouldn't happen, but consequences are pretty bad if we have the lock held and
+	// our caller tries to take the lock again on panic defer, as in processSnapshot.
+	if state == nil || snap == nil || mset.node == nil {
```
Comment (wallyqs) [inline server/jetstream_cluster.go:7919]: do we need to check `mset == nil`?

Comment (neilalexander) [reply server/jetstream_cluster.go:7919]: If `mset == nil`, then `processSnapshot` has already panicked because of trying to take the lock.

**Review by wallyqs — APPROVED** [review]
LGTM

**Review by derekcollison — APPROVED** [review]
LGTM

**Pull request merged**